### ISSUE TITLE
一些疑问

### ISSUE THREAD

**Issue by xxllp** [shenweichen/DeepCTR#3]
这个里面的tf 和keras 模型运行的结果有一定的差别，不知道为啥

我换了一个大一点的数据运行直接就卡了

**Comment by shenweichen** [comment]
tf和keras模型的运行结果不保证完全一致，因为可能一些参数初始化方法，还有随机数种子的设置都会导致不同的~
至于大一点数据卡了是什么情况？batch_size设小一点看看有没有输出，另外大模型刚开始需要一定时间进行初始化。

**Comment by xxllp** [comment]
我用了一个20w *100多字符变量的数据 ，等了20分钟不止了 ，一直没有进入训练的日志。也太慢了，能哪里加速下吗

**Comment by shenweichen** [comment]
是在数据预处理过程中卡住了 还是在模型训练过程中卡住了呢 比如 LabelEncoder的话可能会慢一些，用多进程的方式处理可以更快，另外现在正在开发使用Hash编码的特征接口，就不用一定要把类别特征预处理成连续编码的形式了。

**Comment by xxllp** [comment]
[array([26, 26, 26, ..., 26, 26, 26]), array([21, 21, 21, ..., 21, 21, 21]), array([19, 20, 20, ..., 19, 19, 19]), array([32, 58, 58, ..., 32, 32, 32]), array([28, 28, 28, ..., 28, 28, 28]), array([34, 34, 41, ..., 34, 34, 34]), array([32, 80, 84, ..., 32, 32, 32]), array([28, 28, 28, ..., 28, 28, 28]), array([36, 36, 36, ..., 36, 34, 36]), array([22, 22, 22, ..., 22, 19, 22]), array([32, 62, 56, ..., 32, 34, 32]), array([33, 29, 29, ..., 33, 31, 33]), array([33, 18, 18, ..., 33, 33, 33]), array([27, 14, 14, ..., 27, 27, 27]), array([29, 83, 87, ..., 29, 29, 29]), array([26, 86, 90, ..., 26, 26, 26]), array([29, 79, 83, ..., 29, 37, 29]), array([35, 87, 91, ..., 35, 44, 35]), array([34, 27, 92, ..., 34, 30, 34]), array([37, 15, 63, ..., 37, 47, 37]), array([28, 86, 92, ..., 28, 28, 28]), array([40, 26, 88, ..., 40, 40, 40]), array([23, 83, 83, ..., 23, 33, 23]), array([37, 17, 89, ..., 37, 37, 37]), array([42, 22, 86, ..., 42, 51, 42]), array([44, 76, 80, ..., 44, 18, 44]), array([ 39,  19, 101, ...,  39,  50,  39]), array([40, 21, 78, ..., 40, 48, 40]), array([40, 25, 88, ..., 40, 50, 40]), array([34, 27, 94, ..., 34, 44, 34]), array([43, 34, 71, ..., 43, 53, 43]), array([ 35,  27, 103, ...,  35,  46,  35]), array([41, 77, 81, ..., 41, 65, 41]), array([30, 58, 58, ..., 31, 20, 31]), array([44, 20, 82, ..., 45, 24, 45]), array([36, 13, 92, ..., 36, 16, 36]), array([42, 26, 65, ..., 42, 28, 42]), array([38, 29, 73, ..., 38, 20, 38]), array([32, 62, 62, ..., 32, 14, 35]), array([41, 14, 27, ..., 45, 16, 45]), array([ 34,  88, 102, ...,  35,   2,  35]), array([27, 57, 58, ..., 27, 51, 27]), array([45, 14, 25, ..., 48, 19, 48]), array([43, 31, 75, ..., 43, 24, 43]), array([34, 11, 25, ..., 40, 12, 40]), array([34, 24, 87, ..., 34, 20, 34]), array([ 23,  57,  73, ...,  87, 104,  87]), array([38, 31, 96, ..., 40, 25, 46]), array([31, 19, 78, ..., 31, 24, 31]), array([21, 62, 50, ..., 72, 65, 72]), array([24, 87, 90, ..., 92, 77, 92]), array([32, 62, 61, ..., 70, 42, 70]), array([52, 50, 29, ..., 62, 94, 78]), array([21, 59, 59, ..., 62, 45, 62]), array([14, 47, 33, ..., 62, 86, 62]), array([ 48,  57,  25, ..., 105,  47,  90]), array([54, 34, 64, ..., 74, 94, 88]), array([29, 51, 62, ..., 64, 64, 67]), array([49, 49, 21, ..., 98, 95,  4]), array([21, 49, 26, ..., 71, 11, 72]), array([21, 53, 34, ..., 67, 53, 53]), array([36, 44, 36, ..., 64, 52, 13]), array([52, 60, 21, ..., 77, 52,  2]), array([73, 82, 83, ..., 94, 57, 87]), array([77, 31, 77, ..., 77, 77, 77]), array([72, 58, 72, ..., 66, 65, 66]), array([46, 57, 29, ..., 10, 94, 43]), array([34, 34, 34, ..., 16, 37,  4]), array([25, 25, 39, ..., 60, 76, 76]), array([43, 45, 44, ..., 74, 57, 76]), array([19, 46, 82, ..., 82, 81, 81]), array([15, 14, 16, ..., 41, 42, 42]), array([33, 76, 88, ..., 75, 76, 76]), array([29, 49, 23, ..., 30, 26, 30]), array([10, 49, 49, ..., 54, 55, 62]), array([ 7,  7,  7, ..., 17, 38,  0]), array([53, 53, 53, ..., 65, 52, 65]), array([50, 32, 71, ..., 71, 70, 71]), array([23, 24, 24, ..., 24, 24, 24]), array([35, 35, 39, ..., 12, 70, 86]), array([54, 43, 55, ..., 44, 59, 43]), array([25, 80, 79, ..., 17, 79,  7]), array([72, 64, 72, ..., 57, 69, 68]), array([36, 70, 52, ...,  6, 68,  6]), array([21, 21, 21, ..., 21, 21, 21]), array([20, 27, 20, ..., 30, 18, 20]), array([14, 14, 60, ..., 65, 63, 65]), array([28, 63, 48, ..., 58, 58, 58]), array([18, 54, 54, ..., 57, 54,  2]), array([64, 68, 68, ..., 68, 38, 76]), array([62, 56, 75, ..., 62, 85, 72]), array([23, 51, 23, ..., 26, 51, 13]), array([62, 34, 68, ..., 32, 67, 67]), array([43, 56, 35, ..., 57, 14,  2]), array([47, 31, 24, ..., 93, 93, 93]), array([ 4, 28, 31, ..., 28, 50, 56]), array([29, 29, 29, ..., 29, 29, 29]), array([19,  1, 19, ..., 27, 30, 30]), array([18, 34, 24, ..., 25, 21, 21]), array([27, 19, 64, ..., 55, 55, 59]), array([46, 46, 46, ..., 95, 88, 94]), array([45, 63, 64, ..., 97, 70, 68]), array([30, 32, 32, ..., 32, 25, 34]), array([21, 21, 21, ..., 21, 23, 23]), array([13, 13, 30, ...,  5, 37,  5]), array([22, 59, 22, ..., 59, 60, 11]), array([13, 29, 13, ...,  2, 38,  2]), array([47, 47, 47, ..., 50, 62, 47]), array([20, 24, 20, ..., 20, 29, 20]), array([23,  5, 76, ..., 38, 76, 67]), array([18, 64, 18, ..., 98, 64, 38]), array([40, 34, 56, ..., 65, 25, 41]), array([17, 17, 17, ..., 22, 17,  8]), array([72, 53, 14, ..., 65, 89, 87]), array([39, 68, 77, ..., 61, 77, 61]), array([28, 28, 34, ..., 37, 34, 34]), array([21, 21, 21, ..., 22, 22, 22]), array([31, 23, 23, ..., 26, 27, 31]), array([21, 46, 21, ..., 94, 22, 24]), array([23, 71, 23, ..., 32, 66, 45])]
WARNING:tensorflow:From /home/mlp/python/anaconda3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:1247: calling reduce_sum (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
WARNING:tensorflow:From /home/mlp/python/anaconda3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:1213: calling reduce_max (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
total embed size 2139
WARNING:tensorflow:From /home/mlp/python/anaconda3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:1264: calling reduce_prod (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
WARNING:tensorflow:From /home/mlp/python/anaconda3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:1349: calling reduce_mean (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
Train on 147764 samples, validate on 36941 samples
Epoch 1/10
2018-08-16 15:13:41.790578: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-08-16 15:13:42.010050: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 0 with properties:
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:02:00.0
totalMemory: 10.91GiB freeMemory: 8.04GiB
2018-08-16 15:13:42.221082: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 1 with properties:
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:03:00.0
totalMemory: 10.91GiB freeMemory: 8.21GiB
2018-08-16 15:13:42.221259: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Device peer to peer matrix
2018-08-16 15:13:42.221299: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1126] DMA: 0 1
2018-08-16 15:13:42.221312: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1136] 0:   Y Y
2018-08-16 15:13:42.221321: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1136] 1:   Y Y
2018-08-16 15:13:42.221338: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:02:00.0, compute capability: 6.1)
2018-08-16 15:13:42.221350: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:1) -> (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:03:00.0, compute capability: 6.1)

**Comment by xxllp** [comment]
卡到这里就啥都没了。。

**Comment by xxllp** [comment]
这个还是没初始化完吗

**Comment by xxllp** [comment]
我数据打印结果在上面

**Comment by shenweichen** [comment]
用的是哪个模型？ 样本数有20W 特征有100个field，每个field大概多少维？

**Comment by xxllp** [comment]
data = pd.read_csv("train.csv")

    print(type(data))
    sparse_features =  ['gbdt_leaf_' + str(i) for i in range(120)]
    target = ['label']

    #1.Label Encoding for sparse features,and Normalization for dense fetures
    for feat in sparse_features:
        lbe = LabelEncoder()
        data[feat] = lbe.fit_transform(data[feat])
    #2.count #unique features for each sparse field
    sparse_feature_dim = {feat:data[feat].nunique() for feat in sparse_features}
    #3.generate input data for model
    model_input = [data[feat].values for feat in sparse_feature_dim] #+ [data[target[0]].values]
    print(model_input)
    #4.Define Model,compile and train
    model_lists = [AFM({"sparse":sparse_feature_dim,"dense":[]},).model,
                   DCN({"sparse":sparse_feature_dim,"dense":[]},).model,
                   DeepFM({"sparse":sparse_feature_dim,"dense":[]},).model,
                   MLR({"sparse":sparse_feature_dim,"dense":[]},).model,
                   NFM({"sparse":sparse_feature_dim,"dense":[]},).model]
    for model in model_lists:
        #model = DeepFM({"sparse":sparse_feature_dim,"dense":[]},final_activation='linear').model
        model.compile("adam","binary_crossentropy",metrics=['mse'],)
        history = model.fit(model_input,data[target],
                  batch_size=256,epochs=10,verbose=2,validation_split=0.2,)
        model.predict(model_input)

**Comment by xxllp** [comment]
就是120个字段，

**Comment by xxllp** [comment]
模型我也不知道选哪个好

**Comment by shenweichen** [comment]
sparse_feature_dim 把这个打印出来看下, 然后batch_size设置成1看下

**Comment by xxllp** [comment]
这个dim 不大的，不超过50 。

**Comment by xxllp** [comment]
我吧size设小看下

**Comment by shenweichen** [comment]
`history = model.fit(model_input,data[target],
              batch_size=256,epochs=10,verbose=2,validation_split=0.2,)`这句改成`verbose=1`看看进度条走不走

**Comment by xxllp** [comment]
到那里还是卡了，等等看吧

**Comment by xxllp** [comment]
这种是最坑的，连个报错都没
